src/Shared/Buffers/RefPooledArrayBufferWriter.cs

```diff
@@ -0,0 +1,266 @@
+        }
+
+        // either we throw, or copy the Span<T> data into a heap-allocated array (via pooling), and then return Memory<T>.
+        // For demonstration, that it should not be used with Memory<T> unless done something specific, we throw here.
+        throw new InvalidOperationException("Cannot convert Span<T> to Memory<T>");
+    }
+
+    /// <summary>
+    /// Gets a span representing the available space for writing.
+    /// </summary>
+    /// <param name="sizeHint">A hint about the minimum size needed. Ignored in this implementation as resizing is not performed.</param>
+    /// <returns>A Span&lt;byte&gt; for the available space.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public Span<T> GetSpan(int sizeHint = 0)
+    {
+        ThrowIfDisposed();
+        CheckAndResizeBuffer(sizeHint);
+        return _rentedBuffer is not null
+            ? _rentedBuffer.AsSpan(_index)
+            : _buffer.Slice(_index);
+    }
+
+    public void Advance(uint count)
+        => Advance((int)count);
+
+    /// <summary>
+    /// Advances the write position by the specified count.
+    /// </summary>
+    /// <param name="count">The number of bytes written.</param>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown if count is negative.</exception>
+    /// <exception cref="InvalidOperationException">Thrown if advancing would exceed buffer capacity.</exception>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void Advance(int count)
+    {
+        ThrowIfDisposed();
+        if (count < 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(count), "Count cannot be negative.");
+        }
+
+        var canAdvance = _rentedBuffer is not null
+            ? _index + count < _rentedBuffer.Length
+            : _index + count < _buffer.Length;
+
+        if (!canAdvance)
+        {
+            throw new InvalidOperationException($"Cannot advance past the end of the buffer. Current position: {_index}, Capacity: {_buffer.Length}, Requested advance: {count}.");
+        }
+        _index += count;
+    }
+
+    private void CheckAndResizeBuffer(int sizeHint)
+    {
+        if (sizeHint <= 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(sizeHint), actualValue: sizeHint, $"{nameof(sizeHint)} ('{sizeHint}') must be a non-negative value.");
+        }
+        if (sizeHint == 0)
+        {
+            sizeHint = MinimumBufferSize;
+        }
+
+        if (_rentedBuffer is null)
+        {
+            var bufferSpace = _buffer.Length - _index;
+            if (bufferSpace < sizeHint)
+            {
+                // initial buffer is not enough, we need to start renting from the pool
+                var rentedInitialSize = _buffer.Length + Math.Max(sizeHint, _buffer.Length);
+                _rentedBuffer = ArrayPool<T>.Shared.Rent(rentedInitialSize);
+
+                _buffer.CopyTo(_rentedBuffer);
+                _buffer.Clear();
```

Comment:
Wouldn't it make the rest of the code a lot simpler if we just reassigned `_buffer` here? At this point, we have no more use for the previous `_buffer`, right?

```suggestion
                _buffer = _rentedBuffer.AsSpan();
```